### PR TITLE
DM-14435 Implement the visual display of MOC in either outline or fill style in Firefly

### PR DESCRIPTION
**Pull request by cwang2016**
The development includes,
- load MOC fits table and collect cell NUNIQ numbers from MOC fits. 
- collecting 4 corners for each cell with a NUNIQ number.  
- derive polygon region for each cell and the polygon is represented by the outmost points of the cell at display resolution (sub-divide the cell of lower order into those of higher order and collecting the points from the side between the 4 corners). The polygon points are adjusted as the cell is moved towards the edge of the display in case MOC is shown in 'fill' style. 
- create DrawObject to represent all cells in MOC fits table. 
- create drawing layer for each MOC. 
- display the MOC in either outline or fill style. The  NUNIQ number is optionally shown with the cell. The color and transparency of MOC can be changed in standard way. 
- MOC layer is shown as a HIPS layer with checkbox off. 

test:
- do HIPS search
- check  the MOC plot layer in 'Layers' edit panel.  
   check either 'outline' or 'fill' for MOC display,  change the color from color panel, and check on/off the showing of the nuniq numbers. 
- the cell is highlighted by mouse clicking.

### CODE REVIEW

**Comment by robyww** [comment]
I am not looking closely yet but a quick comment. For SDSS the MOC does not appear to be consistent with the data. I don't know if that is because the MOC is bad or the code is wrong.  You might want to try it in the Aladin application.

**Comment by cwang2016** [comment]
@robyww do you mean the one with title 'SDSS9 color'?

**Comment by cwang2016** [comment]
-fixed the bug that showed only  100 tiles maximally from MOC fits file. 
-some work is made to improve the MOC display performance (speed and accuracy)

**Comment by robyww** [comment]
We should talk about that tomorrow as well.

**Comment by cwang2016** [comment]
Here is some update made to improve the performance of MOC display, 
- load the MOC fits when the MOC layer is checked up to be shown. 
- change the computation and rendering of MOC layer to be out of reduce context in order not to prevent the interaction with the UI.

File: src/firefly/js/visualize/draw/ShapeDataObj.js [hunk 1]
```diff
@@ -1002,33 +1020,256 @@ function drawEllipse(drawObj, ctx, plot, drawParams, onlyAddToPath) {
...
+ * draw polygon - draw outline of the polygon or fill the polygon
+ * @param drawObj
+ * @param ctx
+ * @param plot
+ * @param drawParams
+ * @param onlyAddToPath
+ */
+function drawPolygon(drawObj, ctx,  plot, drawParams, onlyAddToPath) {
+    const {style, color, lineWidth=1} = drawParams;
+    const isMocTile = drawObj.mocInfo;
```
Comment (robyww) [inline src/firefly/js/visualize/draw/ShapeDataObj.js]: We have always tried to keep the drawing code separate from the business logic. For example this file knows how to draw shapes.  It is necessary to have knowledge of MOC at this level.  Can it not stay at the drawing layer level?  I would like you to consider this and see.  If you think it must know about MOC at such a low level the we will live with it but it does break the paradigm

Comment (cwang2016) [reply src/firefly/js/visualize/draw/ShapeDataObj.js]: Good point. The logic can be fixed in generic way.

File: src/firefly/js/drawingLayers/HiPSMOC.js [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+ * @returns {Function}
+ */
+function makeUpdateDeferred(dl, plotId) {
+    const {updateStatusAry} = dl;
+
+    updateStatusAry[plotId].startUpdate();
+    addTask(plotId, updateStatusAry[plotId].updateTaskId);
+
+    const id = window.setInterval( () => {
+        updateMocData(dl, plotId);
```
[the remark this answers is not in the document]
Comment (cwang2016) [reply src/firefly/js/drawingLayers/HiPSMOC.js]: yes, it is in updateDrawLayer.

File: src/firefly/js/drawingLayers/HiPSMOC.js [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+ * @param plotId
+ */
+function updateMocData(dl, plotId) {
+    const {updateStatusAry, mocObj} = dl;
+    const plot = primePlot(visRoot(), plotId);
+    const updateStatus = updateStatusAry[plotId];
+
+     if (isEmpty(updateStatus.newMocObj)) {    // find visible cells first
+        const newMocObj = clone(mocObj);
+        newMocObj.mocGroup = new MocGroup(null, mocObj.mocGroup, plot);
```
[the remark this answers is not in the document]
Comment (cwang2016) [reply src/firefly/js/drawingLayers/HiPSMOC.js]: yes, updated

**Comment by cwang2016** [comment]
Apply the drawing layer optimization method to MOC display in 'outline' style.

**Pull request merged**